dvc/utils/fs.py

```diff
@@ -20,7 +21,7 @@ def get_mtime_and_size(path):
     mtime = os.path.getmtime(path)
 
     if os.path.isdir(path):
-        for root, dirs, files in os.walk(str(path)):
+        for root, dirs, files in dvc_walk(str(path)):
```

Comment:
This will change mtime/size logic for directories. Did you test it with `dvc add directory` that has ignored files in it? What if one of that ignored files changes? It will change mtime of the directory itself, right? Which will make state entry not valid and will make dvc re-compute checksum for it. Should we not account for mtimes and sizes of the directories itself and only rely on files?

Reply:
Summary of private conversation about get_mtime_and_size:
We have a problem when we want to use get_mtime_and_size when it contains .dvcignore-ed paths because:
- we should not take into account ignored  files, because updating ignored file will result in updated directory mtime, which will (for example) cause checksum recalculation, when actual data (not ignored) has not been modified


Straightforward solution for that is to analyse files only, however that causes another problem:
- renaming file under directory will not cause update of this file mtime, but it will cause update of parent directory mtime. So, in such case, analysing files only will not return changed mtime, which is obviously not true.

We need some smarter way to handle get_mtime_and_size.

Reply:
Summary of another private conversation:
- Because of resons mentioned in post above, we need mechanism that will detect changes in dir, but also will be immune to changes performed on "ignored" files.

New way of dealing with this:
1. Let size be calculated as is, because modifications of ignored files will not affect dir size calculations (duirng walk we ignore those files)
2. Develop new method, that will be used to detect changes, effectively replacing mtime update. How to do that? Put all files for given dir in list `[(relpath: mtime)]` and in the end don't extract greatest mtime(as it is done now), but pass this object through checksum method. Thanks to that we will be able to detect changes in directory tree, yet we will not have to worry about update of ignored files updating mtime of parent (and not ignored) directory.